## Data Loading

In [1]:
# import libraries

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
pd_count_points = pd.read_csv('../data/dft_countpoints_region_id_7.csv')

In [3]:
pd_raw_count = pd.read_csv('../data/dft_rawcount_region_id_7.csv')

/var/folders/_z/nz5spzqj2rl11n5lc_k1cwwc0000gn/T/ipykernel_62722/254017237.py:1: DtypeWarning: Columns (11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  pd_raw_count = pd.read_csv('../data/dft_rawcount_region_id_7.csv')


## Data Inspection

#### Basic Inspection of pd_count_points dataset

In [4]:
pd_count_points.sample()

count_point_id  region_id      region_name  local_authority_id  \
3543           81084          7  East of England                 120   

     local_authority_name road_name road_type start_junction_road_name  \
3543                Luton     A5228     Major     B579 - Leagrave Road   

     end_junction_road_name  easting  northing   latitude  longitude  \
3543                     A6   508220    223140  51.896312  -0.428648   

      link_length_km  link_length_miles  
3543             1.3               0.81

In [5]:
pd_count_points.columns

Index(['count_point_id', 'region_id', 'region_name', 'local_authority_id',
       'local_authority_name', 'road_name', 'road_type',
       'start_junction_road_name', 'end_junction_road_name', 'easting',
       'northing', 'latitude', 'longitude', 'link_length_km',
       'link_length_miles'],
      dtype='object')

In [6]:
print(f"The shape for count_points:\n{pd_count_points.shape}")

The shape for count_points:
(4981, 15)


In [7]:
pd_count_points.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4981 entries, 0 to 4980
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   count_point_id            4981 non-null   int64  
 1   region_id                 4981 non-null   int64  
 2   region_name               4981 non-null   object 
 3   local_authority_id        4981 non-null   int64  
 4   local_authority_name      4981 non-null   object 
 5   road_name                 4981 non-null   object 
 6   road_type                 4981 non-null   object 
 7   start_junction_road_name  1867 non-null   object 
 8   end_junction_road_name    1867 non-null   object 
 9   easting                   4981 non-null   int64  
 10  northing                  4981 non-null   int64  
 11  latitude                  4981 non-null   float64
 12  longitude                 4981 non-null   float64
 13  link_length_km            1872 non-null   float64
 14  link_len

#### Basic Inspection of pd_raw_count dataset

In [8]:
pd_raw_count.sample()

count_point_id direction_of_travel  year  count_date  hour  region_id  \
11026            6689                   N  2014  2014-06-12    17          7   

           region_name  local_authority_id local_authority_name road_name  \
11026  East of England                 126              Suffolk     A1117   

       ... buses_and_coaches lgvs hgvs_2_rigid_axle  hgvs_3_rigid_axle  \
11026  ...                 5  101               4.0                  1   

       hgvs_4_or_more_rigid_axle  hgvs_3_or_4_articulated_axle  \
11026                          0                             1   

       hgvs_5_articulated_axle  hgvs_6_articulated_axle  all_hgvs  \
11026                        0                        0       6.0   

       all_motor_vehicles  
11026               794.0  

[1 rows x 32 columns]

In [9]:
pd_raw_count.columns

Index(['count_point_id', 'direction_of_travel', 'year', 'count_date', 'hour',
       'region_id', 'region_name', 'local_authority_id',
       'local_authority_name', 'road_name', 'road_type',
       'start_junction_road_name', 'end_junction_road_name', 'easting',
       'northing', 'latitude', 'longitude', 'link_length_km',
       'link_length_miles', 'pedal_cycles', 'two_wheeled_motor_vehicles',
       'cars_and_taxis', 'buses_and_coaches', 'lgvs', 'hgvs_2_rigid_axle',
       'hgvs_3_rigid_axle', 'hgvs_4_or_more_rigid_axle',
       'hgvs_3_or_4_articulated_axle', 'hgvs_5_articulated_axle',
       'hgvs_6_articulated_axle', 'all_hgvs', 'all_motor_vehicles'],
      dtype='object')

In [10]:
print(f"The shape for pd_raw_count:\n{pd_raw_count.shape}")

The shape for pd_raw_count:
(550056, 32)


In [11]:
pd_raw_count.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550056 entries, 0 to 550055
Data columns (total 32 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   count_point_id                550056 non-null  int64  
 1   direction_of_travel           550056 non-null  object 
 2   year                          550056 non-null  int64  
 3   count_date                    550056 non-null  object 
 4   hour                          550056 non-null  int64  
 5   region_id                     550056 non-null  int64  
 6   region_name                   550056 non-null  object 
 7   local_authority_id            550056 non-null  int64  
 8   local_authority_name          550056 non-null  object 
 9   road_name                     550056 non-null  object 
 10  road_type                     550056 non-null  object 
 11  start_junction_road_name      220716 non-null  object 
 12  end_junction_road_name        220716 non-nul

#### Data Quality checks

- Consistency Check: Ensure that all count_point_id in pd_raw_count are also in pd_count_points

In [12]:
print("\nUnique count_point_id in pd_count_points:", pd_count_points['count_point_id'].nunique())
print("Total rows in pd_count_points:", len(pd_count_points))


Unique count_point_id in pd_count_points: 4981
Total rows in pd_count_points: 4981


In [13]:
# Ensure that all count_point_id in pd_raw_count are also in pd_count_points
inconsistent_ids = pd_raw_count[~pd_raw_count['count_point_id'].isin(pd_count_points['count_point_id'])]
print("\nInconsistent count_point_ids (not found in pd_count_points):")
print(inconsistent_ids)


Inconsistent count_point_ids (not found in pd_count_points):
        count_point_id direction_of_travel  year  count_date  hour  region_id  \
442404          941987                   E  2009  2009-06-08     7          7   
442405          941987                   E  2009  2009-06-08     8          7   
442406          941987                   E  2009  2009-06-08     9          7   
442407          941987                   E  2009  2009-06-08    10          7   
442408          941987                   E  2009  2009-06-08    11          7   
...                ...                 ...   ...         ...   ...        ...   
442543          941987                   W  2014  2014-06-09    14          7   
442544          941987                   W  2014  2014-06-09    15          7   
442545          941987                   W  2014  2014-06-09    16          7   
442546          941987                   W  2014  2014-06-09    17          7   
442547          941987                   W  201

## Data Cleaning

TODO:

   - Check that datatypes and formatting is all correct.
   - Check for duplicates.
   - Check for null (missing) values.
   - additionally: erroneous data

#### Data Formatting

**Notes**

- Transform to datetime: `count_date` 
- Extract Date Components and create new columns for the year, month, day, and day of the week

**Considerations**
- Transform to categorical: county, state, race
- Transform to numerical: count

In [14]:
# Check data types
print("\nData types in pd_count_points:")
print(pd_count_points.dtypes)


Data types in pd_count_points:
count_point_id                int64
region_id                     int64
region_name                  object
local_authority_id            int64
local_authority_name         object
road_name                    object
road_type                    object
start_junction_road_name     object
end_junction_road_name       object
easting                       int64
northing                      int64
latitude                    float64
longitude                   float64
link_length_km              float64
link_length_miles           float64
dtype: object


**Convert Object to datatime datatype**

In [15]:
# conver `count_date` to datetime type
pd_raw_count['count_date'] =  pd.to_datetime(pd_raw_count['count_date'])

In [16]:
# Check data types
print("\nData types in pd_raw_count:")
print(pd_raw_count.dtypes)


Data types in pd_raw_count:
count_point_id                           int64
direction_of_travel                     object
year                                     int64
count_date                      datetime64[ns]
hour                                     int64
region_id                                int64
region_name                             object
local_authority_id                       int64
local_authority_name                    object
road_name                               object
road_type                               object
start_junction_road_name                object
end_junction_road_name                  object
easting                                  int64
northing                                 int64
latitude                               float64
longitude                              float64
link_length_km                         float64
link_length_miles                      float64
pedal_cycles                             int64
two_wheeled_motor_vehicles     

**Create new columns for the year, month, day, and day of the week**

In [17]:
# Extract Date Components 
pd_raw_count['Month'] = pd_raw_count['count_date'].dt.month
pd_raw_count['Day_of_Week'] = pd_raw_count['count_date'].dt.day_name()
pd_raw_count['Day'] = pd_raw_count['count_date'].dt.day

### Check Data Duplicates

**Notes**:

- No duplicates found in either dataset as data already aggrated

In [18]:
pd_count_points.duplicated().value_counts()

False    4981
Name: count, dtype: int64

In [19]:
pd_raw_count.duplicated().value_counts()

False    550056
Name: count, dtype: int64

### Check for Nulls

**Steps**
- Check for Nulls
- Visualise Nulls

In [20]:
# Checking for missing values 
print("Missing values in pd_count_points:")
print((pd_count_points.isna().sum() * 100 / pd_count_points.shape[0]).round(2))

Missing values in pd_count_points:
count_point_id               0.00
region_id                    0.00
region_name                  0.00
local_authority_id           0.00
local_authority_name         0.00
road_name                    0.00
road_type                    0.00
start_junction_road_name    62.52
end_junction_road_name      62.52
easting                      0.00
northing                     0.00
latitude                     0.00
longitude                    0.00
link_length_km              62.42
link_length_miles           62.42
dtype: float64


In [21]:
print("\nMissing values in pd_raw_count:")
print((pd_raw_count.isna().sum() * 100 / pd_raw_count.shape[0]).round(2))


Missing values in pd_raw_count:
count_point_id                   0.00
direction_of_travel              0.00
year                             0.00
count_date                       0.00
hour                             0.00
region_id                        0.00
region_name                      0.00
local_authority_id               0.00
local_authority_name             0.00
road_name                        0.00
road_type                        0.00
start_junction_road_name        59.87
end_junction_road_name          59.87
easting                          0.00
northing                         0.00
latitude                         0.00
longitude                        0.00
link_length_km                  59.83
link_length_miles               59.83
pedal_cycles                     0.00
two_wheeled_motor_vehicles       0.00
cars_and_taxis                   0.00
buses_and_coaches                0.00
lgvs                             0.00
hgvs_2_rigid_axle                0.00
hgvs_3_rigid_axle

**Missing Values Options**

- Impute
- Drop any rows that are affected
- Drop columns that are affected

#### Dealing nulls in pd_count_points 

**Notes** 

Looking at the missing values in start_junction_road_name 59.87 and end_junction_road_name 59.87, we can conclude that base on our project objective and the dataset we have we can either:

- Drop the rows with the missing values 
- Remove Rows with Missing Values: If start_junction_road_name and end_junction_road_name are essential, we can consider removing rows where both are missing, especially if it doesn’t significantly reduce the dataset.
- Fill the missing values with the `road_name`, `latitude` or `longitude` values of the nearest road/count_point

In [22]:
# Remove rows where both 'start_junction_road_name' and 'end_junction_road_name' are missing
pd_count_points = pd_count_points.dropna(subset=['start_junction_road_name', 'end_junction_road_name'], how='all')


In [23]:
pd_count_points.isnull().sum()

count_point_id              0
region_id                   0
region_name                 0
local_authority_id          0
local_authority_name        0
road_name                   0
road_type                   0
start_junction_road_name    0
end_junction_road_name      0
easting                     0
northing                    0
latitude                    0
longitude                   0
link_length_km              0
link_length_miles           0
dtype: int64

**Observation**
- We can observe that by removing both 'start_junction_road_name' and 'end_junction_road_name' are missing, the `link_length_km ` and `link_lenght_km` columns are also sorted. 

#### Dealing nulls in pd_raw_count 

In [25]:
pd_count_points[['end_junction_road_name', 'start_junction_road_name']].isnull().value_counts(normalize=True)

end_junction_road_name  start_junction_road_name
False                   False                       1.0
Name: proportion, dtype: float64

In [28]:
# Remove rows where both 'start_junction_road_name' and 'end_junction_road_name' are missing
pd_raw_count = pd_raw_count.dropna(subset=['start_junction_road_name', 'end_junction_road_name'], how='all')

In [29]:
pd_raw_count.isnull().sum()

count_point_id                  0
direction_of_travel             0
year                            0
count_date                      0
hour                            0
region_id                       0
region_name                     0
local_authority_id              0
local_authority_name            0
road_name                       0
road_type                       0
start_junction_road_name        0
end_junction_road_name          0
easting                         0
northing                        0
latitude                        0
longitude                       0
link_length_km                  0
link_length_miles               0
pedal_cycles                    0
two_wheeled_motor_vehicles      0
cars_and_taxis                  0
buses_and_coaches               0
lgvs                            0
hgvs_2_rigid_axle               1
hgvs_3_rigid_axle               0
hgvs_4_or_more_rigid_axle       0
hgvs_3_or_4_articulated_axle    0
hgvs_5_articulated_axle         0
hgvs_6_articul

In [30]:
pd_raw_count = pd_raw_count.drop(['hgvs_2_rigid_axle', 'all_hgvs'], axis=1)

In [31]:
pd_raw_count.isnull().sum()

count_point_id                  0
direction_of_travel             0
year                            0
count_date                      0
hour                            0
region_id                       0
region_name                     0
local_authority_id              0
local_authority_name            0
road_name                       0
road_type                       0
start_junction_road_name        0
end_junction_road_name          0
easting                         0
northing                        0
latitude                        0
longitude                       0
link_length_km                  0
link_length_miles               0
pedal_cycles                    0
two_wheeled_motor_vehicles      0
cars_and_taxis                  0
buses_and_coaches               0
lgvs                            0
hgvs_3_rigid_axle               0
hgvs_4_or_more_rigid_axle       0
hgvs_3_or_4_articulated_axle    0
hgvs_5_articulated_axle         0
hgvs_6_articulated_axle         0
all_motor_vehi

In [32]:
pd_raw_count = pd_raw_count.drop(['all_motor_vehicles'], axis=1)

In [33]:
pd_raw_count.isnull().sum()

count_point_id                  0
direction_of_travel             0
year                            0
count_date                      0
hour                            0
region_id                       0
region_name                     0
local_authority_id              0
local_authority_name            0
road_name                       0
road_type                       0
start_junction_road_name        0
end_junction_road_name          0
easting                         0
northing                        0
latitude                        0
longitude                       0
link_length_km                  0
link_length_miles               0
pedal_cycles                    0
two_wheeled_motor_vehicles      0
cars_and_taxis                  0
buses_and_coaches               0
lgvs                            0
hgvs_3_rigid_axle               0
hgvs_4_or_more_rigid_axle       0
hgvs_3_or_4_articulated_axle    0
hgvs_5_articulated_axle         0
hgvs_6_articulated_axle         0
Month         

#### Save cleaned data to a new CSV

In [180]:
pd_raw_count.to_csv('../data/pd_raw_count_cleaned.csv', index=False)

In [181]:
pd_count_points.to_csv('../data/pd_count_points_cleaned.csv', index=False)